ktor-core/src/org/jetbrains/ktor/http/Cookies.kt

```diff
@@ -0,0 +1,150 @@
+package org.jetbrains.ktor.http.cookies
+
+import org.jetbrains.ktor.application.*
+import org.jetbrains.ktor.http.*
+import java.net.*
+import java.time.*
+import java.time.temporal.*
+import java.util.*
+
+public enum class CookieEncoding {
+    RAW, DQUOTES, URI_ENCODING, BASE64_ENCODING
+}
+
+public open class RequestCookies(val request: ApplicationRequest) {
+    private val defaultDecoded: Map<String, String> by lazy { decode(CookieEncoding.URI_ENCODING) }
+
+    public fun get(name: String): String? = defaultDecoded[name]
+    public fun decode(encoding: CookieEncoding): Map<String, String> = request.attributes.getOrPut(CookiesKey(encoding)) {
+        parsedRawCookies.mapValues { decodeCookieValue(it.value, encoding) }
+    }
+
+    public open val parsedRawCookies: Map<String, String> by lazy {
+        request.header("Cookie")?.let { parseClientCookiesHeader(it) } ?: emptyMap()
+    }
+}
+
+public fun parseClientCookiesHeader(cookiesHeader: String): Map<String, String> =
+        cookiesHeader.trimStart().split("[;,]\\s*".toRegex())
+                .filter { !it.startsWith("$") && '=' in it }
+                .map { it.split("=") }
+                .toMap({ it[0].trim() }, { it[1] })
+
+public fun renderSetCookieHeader(name: String,
+                                 value: String,
+                                 encoding: CookieEncoding = CookieEncoding.URI_ENCODING,
+                                 maxAge: Int = 0,
+                                 expires: Temporal? = null,
+                                 domain: String = "",
+                                 path: String = "",
+                                 secure: Boolean = false,
+                                 httpOnly: Boolean = false,
+                                 extensions: Map<String, String?> = emptyMap()): String =
+        (listOf(
+                cookiePart(name.assertCookieName(), value, encoding),
+                cookiePartUnencoded("Max-Age", if (maxAge > 0) maxAge else null),
+                cookiePartUnencoded("Expires", expires?.toHttpDateString()),
+                cookiePart("Domain", domain.nullIfEmpty(), CookieEncoding.RAW),
+                cookiePart("Path", path.nullIfEmpty(), CookieEncoding.RAW),
+
+                cookiePartFlag("Secure", secure),
+                cookiePartFlag("HttpOnly", httpOnly)
+        ) + extensions.map {
+            cookiePartExt(it.key.assertCookieName(), it.value, encoding)
+        }).filter { it.isNotEmpty() }
+                .joinToString("; ")
+
+
+public fun ApplicationResponse.setCookie(name: String,
```

Comment:
Interceptors should be able to examine cookies set by handlers and other interceptors. With this solution they will have to intercept headers and decode them back.